bob/src/cluster/quorum.rs

```diff
@@ -32,31 +32,31 @@ impl Quorum {
 
     async fn put_at_least(&self, key: BobKey, data: &BobData) -> Result<(), Error> {
         let mut local_put_ok = 0_usize;
-        let at_least = self.quorum;
+        let mut at_least = self.quorum;
         let mut failed_nodes = Vec::new();
-        let (vdisk_id, disk_path) = self.mapper.get_operation(key);
-        let (tasks, errors) =
-            if let Some(disk_path) = disk_path {
-                debug!("PUT[{}] ~~~PUT TO {} REMOTE NODES AND LOCAL NODE~~~", key, at_least - 1);
-                let (remote_put, local_put) = tokio::join!(
-                    self.put_remote_nodes(key, data, at_least - 1),
-                    put_local_node(&self.backend, key, data, vdisk_id, disk_path),
-                );
-                let (mut remote_tasks, mut errors) = remote_put;
-                if let Err(e) = local_put {
+        let (vdisk_id, disk_paths) = self.mapper.get_operation(key);
+        if let Some(paths) = disk_paths {
+            debug!("disk path is present, try put local");
+            for p in paths {
+                let res = put_local_node(&self.backend, key, data, vdisk_id, p).await;
+                if let Err(e) = res {
                     error!("{}", e);
-                    failed_nodes.push(self.mapper.local_node_name().clone());
-                    debug!("PUT[{}] local failed, put another remote", key);
-                    errors.extend(finish_at_least_handles(&mut remote_tasks, 1).await.into_iter());
+                    if failed_nodes.is_empty() {
+                        failed_nodes.push(self.mapper.local_node_name().to_owned());
+                    }
                 } else {
                     local_put_ok += 1;
+                    at_least -= 1;
                     debug!("PUT[{}] local node put successful", key);
                 }
-                (remote_tasks, errors)
-            } else {
-                debug!("PUT[{}] ~~~PUT TO {} REMOTE NODES~~~", key, at_least);
-                self.put_remote_nodes(key, data, at_least).await
-            };
+            }
+        } else {
+            debug!("skip local put");
+        }
+        debug!("PUT[{}] need at least {} additional puts", key, at_least);
+
+        debug!("PUT[{}] ~~~PUT TO REMOTE NODES~~~", key);
+        let (tasks, errors) = self.put_remote_nodes(key, data, at_least).await;
         let all_count = self.mapper.get_target_nodes_for_key(key).len();
         let remote_ok_count = all_count - errors.len() - tasks.len() - local_put_ok;
         failed_nodes.extend(errors.iter().map(|e| e.node_name().clone()));
```

Comment:
There is a big problem in the line below: `remote_ok_count + local_put_ok >= self.quorum`.
`get_target_nodes_for_key` returns unique remote nodes, and if one of these nodes contains multiple vdisk replicas, then that node should be counted multiple times. 
Example: `quorum = 2`, `number of replicas = 2`. If both replicas of vdisk are on the one node, then `remote_ok_count = 1`, `local_put_ok = 0`. In this case the condition `remote_ok_count + local_put_ok >= self.quorum` will be `false` and this leads to internal error returned from `put_aliens`.

Right now I don't know what is the best way to resolve this problem

Reply:
`VDisk` struct contains both nodes list and replicas list: https://github.com/qoollo/bob/blob/master/bob-common/src/core_types.rs#L45 Therefore, we can check if their lengths are equal or not. If not, then for each request we call a separate function that calculates the number of vdisk replicas for specific node. If equal, obviously, every requests adds 1 to `remote_ok_count`

Reply:
So put to node should return count of written replicas?

Reply:
Adding info to gRPC is the best way to solve the problem, but I think that changing the gRPC protocol within the current version is not worth doing. Right now it is better to treat error as if no one replica was written, and success as all replicas was written. 

I've created an issue for future to fix the problem properly: https://github.com/qoollo/bob/issues/818

Reply:
I've added passing affected replicas to results so we can count actual number of replicas written